glean/src/core/pings/index.ts

```
@@ -44,6 +53,49 @@ class PingType implements CommonPingData {
    *               `ping_info.reason` part of the payload.
    */
   submit(reason?: string): void {
+    // **** Read this before changing the following code! ****
+    //
+    // The Dispatcher does not allow dispatched tasks to await on
+    // other dispatched tasks. Unfortunately, this causes a deadlock.
+    // In order to work around that problem, we kick off validation
+    // right before the actual submission takes place, through another
+    // async function (and not through the dispatcher). We then await
+    // in the dispatched submission task on the related promise, which
+    // will always resolve due to the use of `finally`.
+    if (this.testValidator) {
+      const cleanup = () => {
+        this.resolveTestPromiseFunction = undefined;
+        this.rejectTestPromiseFunction = undefined;
+        this.testValidator = undefined;
+      };
+
+      this.testValidator(reason)
+        .then(() => {
+          // Temporarily store the function and then clean up, so that fast consecutive
+          // calls to `testBeforeNextSubmit` don't fail because things are still set
+          // after the calling promise is resolved.
+          const resolver = this.resolveTestPromiseFunction;
+          cleanup();
+
+          this.internalSubmit(reason, resolver);
+        })
+        .catch(e => {
+          console.error(`There was an error validating "${this.name}" (${reason ?? "no reason"}):`, e);
+
+          // Temporarily store the function and then clean up, so that fast consecutive
+          // calls to `testBeforeNextSubmit` don't fail because things are still set
+          // after the calling promise is resolved.
+          const rejecter = this.rejectTestPromiseFunction;
+          cleanup();
+
+          this.internalSubmit(reason, rejecter);
```

Do we want to submit even if the validator got an error? This is testing code, it is better to raise an exception if a test error happened, otherwise output is misleading i.e. it looks like the tests passed when there are actually error logs around (encountered this confusing scenario many times yesterday when testing this PR).

---

Yes, I think we should submit the ping regardless. I'd rather not change the code too much, we should keep this as close as production as possible, with the only exception that some test code runs before the submission.

The code already rejects/throws (exceptions will cause rejections anyway): see the related test.